---
title: "CCBR1066- Secondary Analysis"
author: "Samantha Sevilla, [CCBR]"
output: html_notebook
editor_options: 
  chunk_output_type: console
output: 
  html_document:
    toc: true
    toc_float: true
---
# Code info
Runs GSEA and ORA with Pipeliner Output for Human data
Developed from code by Vishal Koparde

```{r include=FALSE}
library("tidyverse")
library("stats")
library("affy")
library("DESeq2")
library("edgeR")
library("DT")
library("ggplot2")
library("EnhancedVolcano")
library("msigdbr")
library("enrichplot")
library("clusterProfiler")
library("ggpubr")
library("pheatmap")
```

######################################################################
# ORA/GSEA Functions
######################################################################
```{r functions, include=FALSE}
readdegfile<-function(fn){
  x=read.csv(fn,header=TRUE,sep="\t")
  return(as.data.frame(x))
}
deg2geneList<-function(deg){
  gl=as.data.frame(deg$gsea_ranking_score)
  gl$GN=deg$gene
  colnames(gl)=c("Rank","GeneName")
  gl$absRank=abs(gl$Rank)
  gl=gl[order(gl$absRank,decreasing = TRUE),]
  gl=gl[match(unique(gl$GeneName),gl$GeneName),]
  geneList=gl$Rank
  names(geneList)=as.character(gl$GeneName)
  geneList <- sort(geneList, decreasing = TRUE)
  return(geneList)
}

gsea_plus_plot <- function(gl,t2g,ttl,fn){
  result=GSEA(geneList = gl,TERM2GENE = t2g,eps = 0, pvalueCutoff = 1)
  resultdf=as.data.frame(result)
  write.table(resultdf,file=fn,quote=FALSE,sep="\t",row.names = FALSE,col.names = TRUE)
  
  if(nrow(result)==0){
    print("No sig results for GSEA")
    p1 = ggparagraph(
      paste0("\n\n\n No Sig Results"),
      color = NULL,
      size = 20,
      face = "bold",
      family = NULL,
      lineheight = NULL
    )
  } else{
    p1 = dotplot(result,title=ttl,font.size = 8, showCategory=5, split=".sign")+facet_grid(.~.sign)
  }
  
  return(p1)
}

ora_plus_plot <- function(gl,t2g,ttl,fn){
  result=enricher(gene=gl, TERM2GENE=t2g, pvalueCutoff = 1)
  resultdf=as.data.frame(result)
  write.table(resultdf,file=fn,quote=FALSE,sep="\t",row.names = FALSE,col.names = TRUE)
  
  if(nrow(resultdf)==0){
    print("No sig results for ORA")
    p1 = ggparagraph(
      paste0("\n\n\n No Sig Results"),
      color = NULL,
      size = 20,
      face = "bold",
      family = NULL,
      lineheight = NULL
    )
  } else{
    p1 = dotplot(result,title=ttl,font.size = 8, showCategory=10)
  }
  return(p1)
}

save_plots<-function(p1,p2,p3="",contrast_id,file_name){
  if(p3=="none"){
    mypdf <- cowplot::plot_grid(p1, p2, ncol=1, labels=LETTERS[1:2])
    print(mypdf)
    ggsave(filename = paste(out_dir, contrast_id, "/", contrast_id, file_name, sep=""), 
           height = 8.90, width = 12.80, device = "png", plot = mypdf)
  } else{
    mypdf <- cowplot::plot_grid(p1, p2, p3, ncol=1, labels=LETTERS[1:3])
    print(mypdf)
    ggsave(filename = paste(out_dir, "cluster_", cluster_id, "/", contrast_id, "_",version_id, "/", contrast_id, file_name, sep=""), 
           height = 8.90, width = 12.80, device = "png", plot = mypdf)
  }
}

#for each annotaiton in df, create heatmap for each contrast input
gsea_plus_heatplot<-function(deg_list, out_dir, contrast_list, annotation_df){
  #for each of the annotation types
  for(rownum in rownames(annotation_df)){
    #start deg counter at 1
    counter_deg=1
    
    #create db
    anno_db = msigdbr(species = species,category = annotation_df[rownum,"ID"], 
                             subcategory = annotation_df[rownum,"SUB"]) %>% dplyr::select(gs_name,gene_symbol)
    
    #compare each contrast at this level
    print(paste0("**Heatplots for ", annotation_df[rownum,"file_name"]))
    
    for(contrast_id in contrast_list){
      print(paste0("--",contrast_id))
      # read deg
      deg=readdegfile(deg_list[counter_deg])
      
      #create output dir if needed for pval and NES output
      dir.create(file.path(out_dir), showWarnings = FALSE)
      dir.create(file.path(paste0(out_dir,"pvalue")), showWarnings = FALSE)
      dir.create(file.path(paste0(out_dir,"NES")), showWarnings = FALSE)
    
      #set genelist for ORA
      geneList=deg2geneList(deg)
      
      ## pvalue
      #pull top paths
      tmp = GSEA(geneList = geneList,TERM2GENE = anno_db,eps = 0, pvalueCutoff = 1) #run GSEA
      tmp_df = as.data.frame(tmp@result) #create df from results
      tmp_df = tmp_df[order(tmp_df$pvalue),] #order by pvalue
      tmp_df = tmp_df[1:5,c("ID","pvalue")] #save top 5 pathways
      tmp_df$pvalue = with(tmp_df, -log(pvalue,10)) #recalc pvalue
      tmp_df = tmp_df[order(tmp_df$ID),]
      names(tmp_df)[names(tmp_df) == "pvalue"] <- contrast_id #rename with contrast_id
      rownames(tmp_df) = NULL #remove rownames
      
      #save final df
      if(exists("final_pval")){
        final_pval = merge(final_pval,tmp_df,all=TRUE)
        final_pval[is.na(final_pval)] <- 0
      } else {
        final_pval=tmp_df
      }
      
      ## NES
      #pull top paths
      tmp_df = as.data.frame(tmp@result) #create df from results
      tmp_df = tmp_df[order(tmp_df$NES),] #order by pvalue
      tmp_df = tmp_df[1:5,c("ID","NES")] #save top 5 pathways
      tmp_df = tmp_df[order(tmp_df$ID),]
      names(tmp_df)[names(tmp_df) == "NES"] <- contrast_id #rename with contrast_id
      rownames(tmp_df) = NULL #remove rownames
      
      #save final df
      if(exists("final_nes")){
        final_nes = merge(final_nes,tmp_df,all=TRUE)
        final_nes[is.na(final_nes)] <- 0
      } else {
        final_nes=tmp_df
      }
      
      #increase deg counter
      counter_deg=counter_deg+1
    }

    #once all contrasts are complete, format db and save heatmaps
    if(nrow(final_pval)<2){
      print("Heatmap cannot be generated with less than two sig pathways: pvalues")
      } else{
        print("--Saving heatplot for pvalues")
        #clean final df
        rownames(final_pval)=final_pval$ID #rownames as ID
        final_pval=final_pval[,2:ncol(final_pval)] #remove ID col
        
        #save fig
        p1 = pheatmap(final_pval,
                 color = colorRampPalette(c("white", "blue"))(256),
                 main = paste0("Heatplot of ", annotation_df[rownum,"file_name"], "\n pvalues"))
        ggsave(filename = paste0(out_dir, "pvalue/","heatplot.", annotation_df[rownum,"file_name"], ".png"),
               height= 8.90, width = 12.80, device = "png", plot = p1)
      }
    
    if(nrow(final_nes)<2){
      print("Heatmap cannot be generated with less than two sig pathways: NES")
      } else{
        print("--Saving heatplot for NES")
        #clean final df
        rownames(final_nes)=final_nes$ID #rownames as ID
        final_nes=final_nes[,2:ncol(final_nes)] #remove ID col
        
        #save fig
        p1 = pheatmap(final_nes,
                 color = colorRampPalette(c("blue", "white"))(256),
                 main = paste0("Heatplot of ", annotation_df[rownum,"file_name"], "\n NES"))
        ggsave(filename = paste0(out_dir, "NES/","heatplot.", annotation_df[rownum,"file_name"], ".png"),
               height= 8.90, width = 12.80, device = "png", plot = p1)
      }
    
    #remove prev df
    remove(final_pval,final_nes)
  }
}

main_function<-function(deg_file,out_dir,contrast_id){
  
  # read deg
  deg=readdegfile(deg_file)
  
  #create output dir if needed
  dir.create(file.path(out_dir), showWarnings = FALSE)

  #subset significant genes for ORA
  siggenes=deg[deg$fdr <= p_val & (deg$fc < (-1*fc) | deg$fc > fc),]
  sigGeneList=siggenes$gene
  
  #set genelist for GSEA
  geneList=deg2geneList(deg)
  
  #generate gene lists for C2 with subtypes biocarta, kegg, reactome, wiki
  c2b=msigdbr(species = species, category = "C2", subcategory = "BIOCARTA") %>% 
    dplyr::select(gs_name,gene_symbol)
  
  c2k=msigdbr(species = species, category = "C2", subcategory = "KEGG") %>% 
    dplyr::select(gs_name,gene_symbol)
  
  c2r=msigdbr(species = species, category = "C2", subcategory = "REACTOME") %>%
    dplyr::select(gs_name,gene_symbol)
  
  c2w=msigdbr(species = species, category = "C2", subcategory = "WIKIPATHWAYS") %>%
    dplyr::select(gs_name,gene_symbol)
  
  #generate gene lists for C5 with subtypes MF, BP, CC
  c5gomf=msigdbr(species = species,  category = "C5", subcategory = "GO:MF") %>% dplyr::select(gs_name,gene_symbol)
  c5gobp=msigdbr(species = species,  category = "C5", subcategory = "GO:BP") %>% dplyr::select(gs_name,gene_symbol)
  c5gocc=msigdbr(species = species,  category = "C5", subcategory = "GO:CC") %>% dplyr::select(gs_name,gene_symbol)
  
  ## C2:BIOCARTA
  print("Results for BIOCARTA")
  p1 = gsea_plus_plot(gl=geneList,t2g=c2b,ttl=paste0("GSEA:C2:BIOCARTA - ",contrast_id),
                      fn=paste(out_dir,contrast_id,".c2b.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c2b,ttl=paste0("ORA:C2:BIOCARTA - ",contrast_id),
                     fn=paste(out_dir,contrast_id,".c2b.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c2b.dotplot.png")
  
  ## C2:KEGG
  print("Results for KEGG")
  p1 = gsea_plus_plot(gl=geneList,t2g=c2k,ttl=paste0("GSEA:C2:KEGG - ",contrast_id),
                      fn=paste(out_dir,contrast_id,".c2k.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c2k,ttl=paste0("ORA:C2:KEGG - ",contrast_id),
                     fn=paste(out_dir,contrast_id,".c2k.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c2k.dotplot.png")

  ## C2:REACTOME
  print("Results for REACTOME")
  p1 = gsea_plus_plot(gl=geneList,t2g=c2r,ttl=paste0("GSEA:C2:REACTOME - ",contrast_id),
                      fn=paste(out_dir,contrast_id,".c2r.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c2r,ttl=paste0("ORA:C2:REACTOME - ",contrast_id),
                     fn=paste(out_dir,contrast_id,".c2r.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c2r.dotplot.png")

  ## C2:WIKIPATHWAYS
  print("Results for WIKIPATHWAYS")
  p1 = gsea_plus_plot(gl=geneList,t2g=c2w,ttl=paste0("GSEA:C2:WIKIPATHWAYS - ",contrast_id),
                    fn=paste(out_dir,contrast_id,".c2w.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c2w,ttl=paste0("ORA:C2:WIKIPATHWAYS - ",contrast_id),
                   fn=paste(out_dir,contrast_id,".c2w.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c2w.dotplot.png")

  ## C5:GO:MF
  print("Results for GO:MF")
  p1 = gsea_plus_plot(gl=geneList,t2g=c5gomf,ttl=paste0("GSEA:GO:MF - ",contrast_id),
                    fn=paste(out_dir,contrast_id,".c5gomf.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c5gomf,ttl=paste0("ORA:GO:MF - ",contrast_id),
                   fn=paste(out_dir,contrast_id,".c5gomf.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c5gomf.dotplot.png")

  ## C5:GO:BP
  print("Results for GO:BP")
  p1 = gsea_plus_plot(gl=geneList,t2g=c5gobp,ttl=paste0("GSEA:GO:BP - ",contrast_id),
                    fn=paste(out_dir,contrast_id,".c5gobp.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c5gobp,ttl=paste0("ORA:GO:BP - ",contrast_id),
                   fn=paste(out_dir,contrast_id,".c5gobp.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c5gobp.dotplot.png")

  ## C5:GO:CC
  print("Results for GO:CC")
  p1 = gsea_plus_plot(gl=geneList,t2g=c5gocc,ttl=paste0("GSEA:GO:CC - ",contrast_id),
                    fn=paste(out_dir,contrast_id,".c5gocc.gsea.results.txt",sep=""))
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c5gocc,ttl=paste0("ORA:GO:CC - ",contrast_id),
                   fn=paste(out_dir,contrast_id,".c5gocc.ora.results.txt",sep=""))
  save_plots(p2,p1,contrast_id,".c5gocc.dotplot.png")
}
```

#####################################
# ORA/GSEA Dotplot Run
#####################################
```{r setup, include=FALSE}
#set global variables
p_val = 0.05
species = "Homo sapiens"
fc = 2

#paths
deg_dir = "~/../../Volumes/projects/ccbr1066/deg_012021/"
out_dir = "~/../../Volumes/projects/ccbr1066/analysis/ora_vs_gsea/"

#Mock vs KO
print("**MOCK v KO")
main_function(deg_file = paste0(deg_dir, "DEG_Mock-KO_0.5_0.5/limma_DEG_Mock-KO_all_genes.txt"), 
     out_dir = paste0(out_dir,"Mock.KO/"),
     contrast_id = "Mock.KO")

#WT vs KO
print("**WT v KO")
main_function(deg_file = paste0(deg_dir, "DEG_WT-KO_0.5_0.5/limma_DEG_WT-KO_all_genes.txt"), 
     out_dir = paste0(out_dir,"WT.KO/"),
     contrast_id = "WT.KO")

#Y318A vs KO
print("**Y318A v KO")
main_function(deg_file = paste0(deg_dir, "DEG_Y318A-KO_0.5_0.5/limma_DEG_Y318A-KO_all_genes.txt"), 
     out_dir = paste0(out_dir,"Y318A.KO/"),
     contrast_id = "Y318A.KO")

```

#####################################
# ORA/GSEA Heatmap Run
#####################################
```{r}
contrast_list = c("Mock.KO","WT.KO","Y318A.KO")

deg_list = c(paste0(deg_dir, "DEG_Mock-KO_0.5_0.5/limma_DEG_Mock-KO_all_genes.txt"),
                                paste0(deg_dir, "DEG_WT-KO_0.5_0.5/limma_DEG_WT-KO_all_genes.txt"),
                                paste0(deg_dir, "DEG_Y318A-KO_0.5_0.5/limma_DEG_Y318A-KO_all_genes.txt"))
             
out_dir = "~/../../Volumes/projects/ccbr1066/analysis/ora_vs_gsea/heatplot/"

                       
annotation_df = data.frame(c("C2","C2","C2","C2","C5","C5","C5"),
                           c("BIOCARTA","KEGG","REACTOME","WIKIPATHWAYS","GO:MF","GO:BP","GO:CC"),
                           c("c2b","c2k","c2r","c2w","c5mf","c5bp","c5cc"))
colnames(annotation_df) = c("ID","SUB","file_name")

#create heatplots with pvalues and NES values
gsea_plus_heatplot(contrast_list = contrast_list,
                   deg_list = deg_list,
                   out_dir = out_dir,
                   annotation_df = annotation_df)

```

#####################################
# Merge DEG for KMeans
#####################################
```{r}
deg_dir = "~/../../Volumes/projects/ccbr1066/deg_012021/"

#pull in degs
deg_list = c(paste0(deg_dir, "DEG_Mock-KO_0.5_0.5/limma_DEG_Mock-KO_all_genes.txt"),
                                paste0(deg_dir, "DEG_WT-KO_0.5_0.5/limma_DEG_WT-KO_all_genes.txt"),
                                paste0(deg_dir, "DEG_Y318A-KO_0.5_0.5/limma_DEG_Y318A-KO_all_genes.txt"))
contrast_list = c("Mock.KO","WT.KO","Y318A.KO")

#created merged df of all degs
counter=1
if(exists("merged_df")){
  remove(merged_df)
}
for (deg in deg_list){
  tmp_df = read.csv(deg,sep="\t")[,c("ensid_gene","pvalue","fdr")]
  
  #separate eid and gene
  tmp_df = tmp_df %>%
    separate(ensid_gene, c("eid_complete", "gene"), "\\|")

  #separate eid and version
  tmp_df$eid_tmp = tmp_df$eid_complete
  tmp_df = tmp_df %>%
    separate(eid_tmp, c("eid_partial", "version"), "\\.")

  tmp_df$contrast = contrast_list[counter]
  counter=counter+1
  
  #save final df
  if(exists("merged_df")){
      merged_df = merge(merged_df,tmp_df,all=TRUE)
    } else {
      merged_df=tmp_df
    }
}

head(merged_df)
remove(tmp_df)

#generate kmeans clusters at http://bioinformatics.sdstate.edu/idep/
#input /data/CCBR/projects/ccbr1066/deg_list/merged_deg_limma_012021.txt
#output K means data buttton, saved to /data/CCBR/projects/ccbr1066/analysis/kmeans/Kmeans_bygroup.csv
```

# KMeans functions
```{r}
#functions
kmeans_data_prep<-function(kmeans.path,cluster.group,df.in,version.id,out.dir){
  
  #read in kmeans data
  km_df = read.csv(kmeans.path) 
  km_df = subset(km_df,Cluster==cluster.group)
  km_df = km_df[,c("Eid","Symbol","MockKOfc","WTKOfc","Y318AKOfc")]
  km_df[is.na(km_df)] <- ""
  
  #add variant info, pvalue to keg output
  for(rowid in rownames(km_df)){
    #pull ids from the merged df either the Ensembl with (complete) or without (partial) version number
    search_eid = subset(df.in,eid_complete==km_df[rowid,"Eid"])$gene[1]
    search_eid2 = subset(df.in,eid_partial==km_df[rowid,"Eid"])$gene[1]
    
    #set lookup column
    if(is.na(search_eid)){
      lookup_key = "eid_partial"
    } else{
      lookup_key = "eid_complete"
    }
    
    #add variant id 
    km_df[rowid,"Variant"] = subset(df.in,get(lookup_key)==km_df[rowid,"Eid"])$gene[1]
      
    #find pvalues for each comparison
    contrast_list = c("Mock.KO","WT.KO","Y318A.KO")
    
    for (contrast_id in contrast_list){
      pvalue = subset(df.in, get(lookup_key)==km_df[rowid,"Eid"] & contrast==contrast_id)$pvalue
      fdr = subset(df.in, get(lookup_key)==km_df[rowid,"Eid"] & contrast==contrast_id)$fdr
  
      #if there is a pvalue, add it to the df
      if(length(pvalue)!=0){
        km_df[rowid,paste0(contrast_id,".p")] = pvalue
      }
      #if there is an fdr value, add it to the df
      if(length(fdr)!=0){
        km_df[rowid,paste0(contrast_id,".fdr")] = fdr
      }
    }
  }

  #cleanup symbol / variant
  for(rowid in rownames(km_df)){
    #if the symbol and variant are the same, remove variant
    if(km_df[rowid,"Symbol"]==km_df[rowid,"Variant"]){
      km_df[rowid,"Variant"] = ""
      
      #if there is no symbol, add variant to symbol
    } else if (km_df[rowid,"Symbol"]==""){
      km_df[rowid,"Symbol"] = km_df[rowid,"Variant"]
      km_df[rowid,"Variant"] = ""
    }
  }
  
  #fix pvalue na
  km_df[is.na(km_df)] <- "1"
  
  #rename headers
  km_df = km_df %>%
    rename(
      Mock.KO.fc = MockKOfc,
      WT.KO.fc = WTKOfc,
      Y318A.KO.fc = Y318AKOfc,
      Gene_Transcript = Symbol
    )
  
  #reorder columns
  km_df_out = km_df[,c("Eid","Gene_Transcript","Variant",
                             "Mock.KO.fc","Mock.KO.p","Mock.KO.fdr",
                             "WT.KO.fc","WT.KO.p","WT.KO.fdr",
                             "Y318A.KO.fc","Y318A.KO.p","Y318A.KO.fdr")]
  #subset sig genes in all comparisons
  km_df_sig = subset(km_df_out,Mock.KO.p<0.05 & WT.KO.p < 0.05 & Y318A.KO.p < 0.05)
  
  #save files
  out_path = paste0(out.dir,"cluster_",cluster.group)
  print(paste0("Saving output to ",out_path))
  dir.create(file.path(out_path), showWarnings = FALSE)
  
  write.csv(km_df_out,paste0(out_path,"/kmeans_stats_all_",version.id,".csv"))
  write.csv(km_df_sig,paste0(out_path,"/kmeans_stats_sig_",version.id,".csv"))
  
  #return df
  return(km_df_out)
}

ora_function<-function(deg.file,out.dir,contrast.id){
  
  # read deg
  deg=deg.file
  
  #create output dir if needed
  dir.create(file.path(out.dir), showWarnings = FALSE)

  #subset significant genes for GSEA
  siggenes=deg[deg$fdr <= p_val & (deg$fc < (-1*fc) | deg$fc > fc),]
  sigGeneList=siggenes$gene
  
  #generate gene lists for C2 with subtypes biocarta, kegg, reactome, wiki
  c2b=msigdbr(species = species, category = "C2", subcategory = "BIOCARTA") %>% 
    dplyr::select(gs_name,gene_symbol)
  
  c2k=msigdbr(species = species, category = "C2", subcategory = "KEGG") %>% 
    dplyr::select(gs_name,gene_symbol)
  
  c2r=msigdbr(species = species, category = "C2", subcategory = "REACTOME") %>%
    dplyr::select(gs_name,gene_symbol)
  
  c2w=msigdbr(species = species, category = "C2", subcategory = "WIKIPATHWAYS") %>%
    dplyr::select(gs_name,gene_symbol)
  
  #generate gene lists for C5 with subtypes MF, BP, CC
  c5gomf=msigdbr(species = species,  category = "C5", subcategory = "GO:MF") %>% dplyr::select(gs_name,gene_symbol)
  c5gobp=msigdbr(species = species,  category = "C5", subcategory = "GO:BP") %>% dplyr::select(gs_name,gene_symbol)
  c5gocc=msigdbr(species = species,  category = "C5", subcategory = "GO:CC") %>% dplyr::select(gs_name,gene_symbol)
  
  ## C2:BIOCARTA
  print("Results for BIOCARTA")
  p1 = ora_plus_plot(gl=sigGeneList,t2g=c2b,ttl=paste0("ORA:C2:BIOCARTA - ",contrast.id),
                     fn=paste(out.dir,contrast.id,".c2b.ora.results.txt",sep=""))

  ## C2:KEGG
  print("Results for KEGG")
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c2k,ttl=paste0("ORA:C2:KEGG - ",contrast.id),
                     fn=paste(out.dir,contrast.id,".c2k.ora.results.txt",sep=""))

  ## C2:REACTOME
  print("Results for REACTOME")
  p3 = ora_plus_plot(gl=sigGeneList,t2g=c2r,ttl=paste0("ORA:C2:REACTOME - ",contrast.id),
                     fn=paste(out.dir,contrast.id,".c2r.ora.results.txt",sep=""))

  save_plots(p1,p2,p3,contrast.id,".c2bkr.dotplot.png")
  
  ## C5:GO:MF
  print("Results for GO:MF")
  p1 = ora_plus_plot(gl=sigGeneList,t2g=c5gomf,ttl=paste0("ORA:GO:MF - ",contrast.id),
                   fn=paste(out.dir,contrast.id,".c5gomf.ora.results.txt",sep=""))

  ## C5:GO:BP
  print("Results for GO:BP")
  p2 = ora_plus_plot(gl=sigGeneList,t2g=c5gobp,ttl=paste0("ORA:GO:BP - ",contrast.id),
                   fn=paste(out.dir,contrast.id,".c5gobp.ora.results.txt",sep=""))

  ## C5:GO:CC
  print("Results for GO:CC")
  p3 = ora_plus_plot(gl=sigGeneList,t2g=c5gocc,ttl=paste0("ORA:GO:CC - ",contrast.id),
                   fn=paste(out.dir,contrast.id,".c5gocc.ora.results.txt",sep=""))

  save_plots(p1,p2,p3,contrast.id,".c2go.dotplot.png")

}
```


#####################################
# ORA on KMeans clusters
#####################################
```{r}
#user parameters
version_id = "v91"
p_val = 0.05
species = "Homo sapiens"
fc = 2
out_dir = "~/../../Volumes/projects/ccbr1066/analysis/kmeans/"

#contrast lists
cluster_list=c("A","B")
contrast_list=c("Mock.KO","Y318A.KO","WT.KO")

#pull kmeans
kmeans_input = paste0("~/../../Volumes/projects/ccbr1066/analysis/kmeans/Kmeans_bygroup_",version_id,".csv")

#for each cluster
for (cluster_id in cluster_list){
  #merge kmeans with degs
  km_means_out = kmeans_data_prep(kmeans_input,cluster_id,merged_df,version_id,out_dir)
  head(km_means_out)
  
  #run ORA plots for each contrast
  for (contrast in contrast_list){
    
    #subset df for specific contrast
    df_in = km_means_out[,c("Gene_Transcript",paste0(contrast,".fc"),paste0(contrast,".p"),paste0(contrast,".fdr"))]
    head(df_in)
    
    #rename subset to match expected input into ora
    df_in = df_in %>%
      rename(
        fc = paste0(contrast,".fc"),
        pvalue = paste0(contrast,".p"),
        fdr = paste0(contrast,".fdr"),
        gene = Gene_Transcript
      )
  
    print("**************")
    print(contrast)
    ora_function(deg.file = df_in, 
       out.dir = paste0(out_dir,"cluster_",cluster_id,"/",contrast,"_",version_id,"/"),
       contrast.id = contrast)
  }
}

```

